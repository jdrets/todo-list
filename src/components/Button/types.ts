export interface ButtonTypes {
  onClick?: () => void
  variant?: string
  hierarchy?: string
  children: any // TODO: add correct type
}
