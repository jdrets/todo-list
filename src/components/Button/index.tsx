import React, { FunctionComponent } from 'react'

import { ButtonTypes } from './types'
import {
  ButtonWrapper
} from './styles'

const Button: FunctionComponent<ButtonTypes> = ({
  children,
  variant,
  onClick,
  disabled
}) => {
  return (
    <ButtonWrapper
      onClick={onClick}
      variant={variant}
      disabled={disabled}
    >
      {children}
    </ButtonWrapper>
  )
}

export default Button
