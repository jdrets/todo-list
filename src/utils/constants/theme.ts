const palette = {
  purpleDark: '#4B00BC',
  purple: '#6600FF',
  purpleMedium: '#8533FF',
  purpleLight: '#E3D0FF',
  green: '#59BC6F',
  greenLight: '#E6F5E2',
  red: '#EB5757',
  redLight: '#FCE4E4',
  yellow: '#DD9900',
  yellowLight: '#FEF1D8',
  white: '#FFFFFF',
  black: '#000000',
  gray: '#898989',
  grayLight: '#E0E0E0'
}

const theme = {
  colors: {
    primaryColor: palette.purple,
    primaryColorMedium: palette.purpleMedium,
    primaryColorLight: palette.purpleLight,
    primaryColorDark: palette.purpleDark,
    ...palette
  },
  layout: {
    breakpoints: {
      medium: '768px'
    },
    mobile: {
      width: '100%',
      padding: '0 24px'
    },
    desktop: {
      width: '1050px'
    }
  }
}

export default theme
