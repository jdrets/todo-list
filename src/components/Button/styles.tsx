import styled from 'styled-components'

import getBackgroundColor from '../../utils/functions/getBackgroundColor'
import getColor from '../../utils/functions/getColor'
import theme from '../../utils/constants/theme'
import { ButtonTypes } from './types'

export const ButtonWrapper = styled.div<ButtonTypes>`
  background-color: ${props => props.disabled ? theme.colors.grayLight : getBackgroundColor(props.variant)};
  border-radius: 24px;
  color: ${props => getColor(props.variant)};
  cursor: pointer;
  pointer-events: ${props => props.disabled && 'none'};
  display: inline-block;
  font-weight: 700;
  padding: 12px 24px;
  transition: 0.3s;
  
  &:hover {
    opacity: .8;
    transition: 0.3s;
  }
`
