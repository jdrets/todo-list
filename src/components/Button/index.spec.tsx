import React from 'react'
import { ThemeProvider } from 'styled-components'

import { render, fireEvent } from '@testing-library/react'

import Button from '.'
import theme from '../../utils/constants/theme'

const mock = {
  onClick: jest.fn()
}

const setup = (mockData = {}) => (
  render(
    <ThemeProvider theme={theme}>
      <Button {...mock} {...mockData}>
        Button
      </Button>
    </ThemeProvider>
  )
)

describe('<Button />', () => {
  test('should match with snapshot', () => {
    const { container } = setup()

    expect(container).toMatchSnapshot()
  })

  test('should call onClick when button is clicked', () => {
    const onClickMock = jest.fn()
    const { getByText } = setup({
      onClick: onClickMock
    })
    const button = getByText('Button')

    fireEvent.click(button)

    expect(onClickMock).toHaveBeenCalled()
  })

  test('should render disable button', () => {
    const { container } = setup({
      disabled: true
    })

    expect(container).toMatchSnapshot()
  })
})
