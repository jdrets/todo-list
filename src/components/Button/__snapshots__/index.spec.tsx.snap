// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match with snapshot 1`] = `
<div>
  <div
    class="sc-bdnylx fhbCOd"
  >
    Button
  </div>
</div>
`;

exports[`<Button /> should render disable button 1`] = `
<div>
  <div
    class="sc-bdnylx drddMu"
    disabled=""
  >
    Button
  </div>
</div>
`;
